core/src/main/scala/pureconfig/DerivedWriters.scala

```diff
@@ -108,8 +119,23 @@ trait DerivedWriters1 {
     }
   }
 
+  // Auxiliary type to make sure we always return `ConfigList` when writing HLists.
+  private[pureconfig] trait HListConfigWriter[T <: HList] extends ConfigWriter[T] {
```

Comment:
At this point, I think that we should drop this hlist-specific trait and turn the two readers below into pure `ConfigReader` instances, even if we have to return an additional failure if an unexpected config type is returned from a tail writer. I think it would be strange as an user if I wanted to provide my own writer for a concrete HList (e.g. `String :: String :: HNil`) and realize that it can't be used as part of a larger HList even if it generates a `ConfigList`.

The concept of restricting the types of configs read/written by some readers/writers in the type system is very interesting, though. We could think of a generic way (via shapeless tags, maybe) to build writers with the guarantee that all generated config values are of a given type, increasing the type-safety of combining them and maybe even serving as a base for generating config schemas.

I'm not too strong in removing this right now, so feel free to leave it as it is :)